### PR TITLE
Fix multiple issues

### LINKED ISSUE #61: Does not support pure Dart packages
Hi thanks for the code generator! However, I need it in pure Dart packages (as well as Flutter packages), but the generator generates code with `import 'package:flutter/widgets.dart';`, which is impossible for dart packages.

Talking about #63, maybe we can create two supportive packages (instead of one), one for dart, one for flutter-specific features.

### LINKED ISSUE #62: Cannot be used in multiple packages
Hi thanks for the lib! Seems that it cannot be used in multiple packages. For example, my app is a package named `my_app`, and it depends on package `my_library`. I want to do i18n for both `my_app` and `my_library`. However, currently it generates two distinct i18n files, so I have to set up them one by one. It would be best if I only need to set up once globally, which will be much better DX.

#63 may solve this problem.

### LINKED ISSUE #63: Extract common code into a support package
HI thanks for the lib! Currently, the `strings.g.dart` contains some infrastructure code that is indeed constant across different packages. Therefore, it would be great if they can be extracted. Take my package (https://github.com/fzyzcjy/flutter_rust_bridge) as an example, I extract those common things into a support package like https://github.com/fzyzcjy/flutter_rust_bridge/blob/9245c9e9f4ae7680f47f0acc45cafb53f5001782/frb_dart/lib/src/basic.dart#L17.

### LINKED ISSUE #64: Linter errors such as "Missing type arguments for map literal."
## 1.

image

Missing type arguments for map literal.  Try adding an explicit type, or remove implicit-dynamic from your analysis options file.

Solution is quite simple: Change it to

```
extension on _StringsEn {
  Map<String, dynamic> _buildFlatMap() {
    return <String, dynamic>{
...
    };
  }
}
```

## 2.

  late final Map<String, dynamic> _flatMap = _buildFlatMap();

Has

Annotate overridden members.

## 3.

```
  // ignore: unused_field
  @override
  final _StringsEn _root;
```

still has unused_field warning, because the "//ignore" is *not* placed *directly* above the line of "_root" variable.

### LINKED ISSUE #66: Let `_t` be a getter instead of a variable
Otherwise, state is redundant, making it easier to introduce bugs

### LINKED ISSUE #67: Remove dependency on `Flutter` for the *code generator*, such that it can be used in pure Dart packages without introducing transitive Flutter dependencies

### LINKED ISSUE #68: RichText/TextSpan support
## Proposal

i18n json:

```
{"myContent(rich)": "In order to see it, you can click [here](NavigateSpanBuilder)."}
```

i18n zh json:

```
{"myContent(rich)": "[点这里](NavigateSpanBuilder)查看。"}
```

generated code:

```
TextSpan myContent({required InlineSpan Function(String) navigateSpanBuilder}) => TextSpan(children: [
  TextSpan(text: 'In order to see it, you can click '),
  navigateSpanBuilder('here'),
  TextSpan(text: '.'),
]);

// zh version is similar, omitted
```

usage:

```
Text.rich(t.myContent(navigateSpanBuilder: (text) => TextSpan(text: text, style: TextStyle(color: Colors.blue), recognizer: TapRecognizer()..onTap=(){...})
```

### PR DESCRIPTION
**Pull request by fzyzcjy**
Close #61 
Close #62 
Close #63 
Close #64
Close #66
Close #67 
Close #68

### CODE REVIEW

File: fast_i18n_flutter/lib/fast_i18n_flutter.dart [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      locale: locale,
+      child: widget.child,
+    );
+  }
+}
+
+class _InheritedLocaleData extends InheritedWidget {
+  final BaseAppLocale locale;
+  Locale get flutterLocale => locale.flutterLocale; // shortcut
+  final _StringsZh translations; // store translations to avoid switch call
```
Comment (Tienisto) [inline fast_i18n_flutter/lib/fast_i18n_flutter.dart:54]: This is difficult to abstract :)

File: lib/src/generator/generate_header.dart [hunk 2]
```diff
@@ -215,11 +216,28 @@ void _generateEnum({
   buffer.writeln(
       '/// - if (LocaleSettings.currentLocale == $baseLocaleEnumConstant) // locale check');
 
-  buffer.writeln('enum $enumName {');
+  buffer.writeln('class $enumName extends BaseAppLocale {');
```
Comment (Tienisto) [inline lib/src/generator/generate_header.dart:219]: Do we still have exhausted `switch` calls?

**Review by Tienisto — COMMENTED** [review]
Lot's of work have been done there!

Some "simple" improvements can be extracted into one cleanup-PR though.
I need to think about some of the breaking changes :)

**Comment by Tienisto** [comment]
Specifically, #64 and #66 can be extracted.

**Comment by fzyzcjy** [comment]
I have not finished everything yet (so it is a draft) :) Will remove the "draft" flag once done.

**Comment by fzyzcjy** [comment]
I know this can be separated into a few PRs, but I need it in my own project, so I commit everything into master branch. Anyway the features themselves will be there.

**Comment by fzyzcjy** [comment]
Ready for review now.

File: fast_i18n_flutter/lib/fast_i18n_flutter.dart (hunk 1 above)
Comment (fzyzcjy) [reply fast_i18n_flutter/lib/fast_i18n_flutter.dart]: Already done (with some modifications though)

File: lib/src/generator/generate_header.dart (hunk 2 above)
Comment (fzyzcjy) [reply lib/src/generator/generate_header.dart:220]: See the new code

**Comment by Tienisto** [comment]
How do I make this run? :D

**Comment by Tienisto** [comment]
Can you explain me how this will fix #62 ?
If you have multiple `strings.g.dart` files, then you have multiple base translations. This should be still the case?

**Comment by Tienisto** [comment]
Okay, I think I understand #62 now.
Basically, you get the `BaseAppLocale` instance from a single source, but you get different translations via the extension `translations` which returns different translations depending on the importing file

**Comment by fzyzcjy** [comment]
> How do I make this run? :D

I run it well on my local machine. Just run code generator. And add dependency of fast_i18n_dart and fast_i18n_flutter.

**Comment by fzyzcjy** [comment]
Thanks for the reply.

> The drawback is that you need to remove the enum. Which is one of (my) favorite features.
Actually, the enum provides a more intelligent linting which is missing now.
> Since the issue of https://github.com/Tienisto/flutter-fast-i18n/issues/62 is about synchronizing locale accross multiple packages, you can store the locale into some global state.

I agree. What about this: Store a BaseAppLocale (not a String - too weak type). And then use the method you proposed to get `t`.

> It's only https://github.com/Tienisto/flutter-fast-i18n/issues/63 which I don't agree 100%

Well, most packages do this separation. For example, json_serializable, mobx, freezed... all big names. Almost all code generator packages do so.

Without extracting support features into a separate package, the code is duplicated across packages, and is hard for other features.

**Comment by Tienisto** [comment]
> I agree. What about this: Store a BaseAppLocale (not a String - too weak type). And then use the method you proposed to get `t`.

The extension method `BaseAppLocale.translations` is fine, there should be some kind of an enum though. Otherwise there will be unnecessary breaking changes (i.e. exhausted switch/if/else are not exhausted anymore).

So basically, (1) `AppLocale` should remain as an enum, (2) Storing the global locale is okay, I agree with your proposed `BaseAppLocale` as a stronger type. How do you want to name it? `BaseAppLocale` seems odd when there is no inheritance. Maybe `AppLocaleId` or `LocaleId`?

Then you can get the `LocaleId` via `LocaleSettings.currentLocale` (the generic is not needed anymore). And from there you get the translations via `LocaleId.translations` (extension method of `strings.g.dart`).

Regarding separate packages: Agreed :)

**Comment by Tienisto** [comment]
I think, we can make this more performant for single-package users :)

`LocaleSettings` should take a constructor parameter `bool useGlobalLocale` (default `false`).
If true, then all translation access will be going through your global `_currLocale` variable.
If false, then `LocaleSettings` should use keep the track of the current locale by storing the enum.

**Comment by Tienisto** [comment]
> Single-source-of-truth is proven to be a good practice and reduce bugs. So maybe we do not need `_useGlobalLocale`?

Agreed

**Comment by fzyzcjy** [comment]
Ready for review now

**Comment by fzyzcjy** [comment]
Well, if not merged immediately, I will implement more features, including #67 and #68 on *this* branch (since it is the master branch of my fork and I need to use the features I implemented immediately in my project)

**Comment by fzyzcjy** [comment]
#68 done. Basicially it is like:

```
// yaml, but json is similar
hello(rich): Hi $username, please ${underline(go to this link)} to see the news
```

```
// generated code
	InlineSpan hello({required Object username, required InlineSpanBuilder underline}) => TextSpan(children: [
		TextSpan(text: 'Hi '),
		username,
		TextSpan(text: ', please '),
		underline('go to this link'),
		TextSpan(text: ' to see the news'),
	]);
```

**Comment by Tienisto** [comment]
Thanks! I will release a dev version soon

**Comment by fzyzcjy** [comment]
You are welcome!

**Pull request merged**